src/main/java/games/strategy/engine/lobby/server/login/RsaAuthenticator.java

```diff
@@ -0,0 +1,183 @@
+    return publicKey;
+  }
+
+  private static KeyPair generateKeypair() {
+    try {
+      final KeyPairGenerator keyGen = KeyPairGenerator.getInstance(RSA);
+      keyGen.initialize(4096);
+      return keyGen.generateKeyPair();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(RSA + " is an invalid algorithm!", e);
+    }
+  }
+
+  private static String decryptPassword(final String encryptedPassword, final String publicKey,
+      final Function<String, String> function) {
+    Preconditions.checkNotNull(encryptedPassword);
+    final PrivateKey privateKey = rsaKeyMap.get(publicKey);
+    if (privateKey == null) {
+      return "Login timeout, try again!";
+    }
+    try {
+      final Cipher cipher = Cipher.getInstance(RSA_ECB_OAEPP);
+      cipher.init(Cipher.DECRYPT_MODE, privateKey);
+      return function
+          .apply(new String(cipher.doFinal(Base64.getDecoder().decode(encryptedPassword)), StandardCharsets.UTF_8));
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+      throw new IllegalStateException(e);
+    } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException e) {
+      return e.getMessage();
+    } finally {
+      rsaKeyMap.remove(publicKey);
+    }
+  }
+
+  /**
+   * Attempts to decrypt the given password using the challenge and response parameters.
+   */
+  public static String decryptPassword(final Map<String, String> challenge, final Map<String, String> response,
+      final Function<String, String> successFullDecryptionAction) {
+    return decryptPassword(response.get(ENCRYPTED_PASSWORD_KEY), challenge.get(RSA_PUBLIC_KEY),
+        successFullDecryptionAction);
+  }
+
+
+  @VisibleForTesting
+  static String encryptPassword(final String publicKeyString, final String password) {
+    try {
+      final PublicKey publicKey =
+          KeyFactory.getInstance(RSA)
+              .generatePublic(new X509EncodedKeySpec(Base64.getDecoder().decode(publicKeyString)));
+      final Cipher cipher = Cipher.getInstance(RSA_ECB_OAEPP);
+      cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+      return Base64.getEncoder().encodeToString(cipher.doFinal(Util.sha512(password).getBytes(StandardCharsets.UTF_8)));
+    } catch (GeneralSecurityException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * This method adds the encrypted password (using the specified public key and password)
+   * to the specified response map.
+   */
+  public static void appendEncryptedPassword(
+      final Map<String, String> response,
+      final Map<String, String> challenge,
+      final String password) {
+    response.put(ENCRYPTED_PASSWORD_KEY, encryptPassword(challenge.get(RSA_PUBLIC_KEY), password));
+  }
+  
+  @VisibleForTesting
+  static void setTimeout(final long amount, final TimeUnit timeUnit) {
+    rsaKeyMap.setTimeout(amount, timeUnit);
+  }
+
+  private static class TimeoutKeyMap extends HashMap<String, PrivateKey> {
```

Comment:
After leaving the comments below, I figured someone must have had a similar need for a class like this.  Sure enough, Guava has a [CacheBuilder](http://google.github.io/guava/releases/19.0/api/docs/com/google/common/cache/CacheBuilder.html) that supports time-based expiration of entries.  It might be best to use that class instead of rolling our own.